pkg/gds/admin/v2/csurf.go

```diff
@@ -0,0 +1,100 @@
+package admin
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+)
+
+// Parameters and names for double-cookie submit CSRF protection
+const (
+	CSRFCookie          = "csrf_token"
+	CSRFReferenceCookie = "csrf_reference_token"
+	CSRFHeader          = "X-CSRF-TOKEN"
+)
+
+// DoubleCookie is Cross Site Request Forgery (CSRF/XSRF) protection middleware that
+// checks the presence of an X-CSRF-TOKEN header containing a cryptographically random
+// token that matches a token contained in the CSRF-TOKEN cookie in the request.
+// Because of the same-origin policy, an attacker cannot access the cookies or scripts
+// of the safe site therefore it will not be able to guess what token to put in the
+// request, thereby protecting from CSRF attacks. Note that Double Cookie CSRF
+// protection requires TLS to prevent MITM attacks.
+func DoubleCookie() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		cookie, err := c.Cookie(CSRFReferenceCookie)
```

Comment:
It is reading the http Request to find a `Cookie` header with the specified name (`csrf_reference_token`) and parsing it into an `*http.Cookie` struct; validating the correctness of the cookie with respect to it's params (such as `Secure`, `MaxAge`, etc.) then returning its value as a string or an error if the Cookie is not found or not valid. If valid, the value of `cookie` should be the token the server specified in `SetCookie` during Authenticate or Reauthenticate. 

I this does bring up an important point - authentication middleware should happen before double cookie middleware since "please login" is a better error then "invalid csrf token".